eth/db/storage.py

```diff
@@ -0,0 +1,470 @@
+            self._trie_nodes_batch = BatchDB(self._db, read_through_deletes=True)
+
+        if self._write_trie is None:
+            storage_root = self._account_db.get_storage_root(self._address)
+            batched_db = self._trie_nodes_batch
+            self._write_trie = HexaryTrie(batched_db, root_hash=storage_root, prune=True)
+            if self._starting_root_hash is None:
+                self._starting_root_hash = storage_root
+
+        return self._write_trie
+
+    def _get_read_trie(self) -> HexaryTrie:
+        storage_root = self._account_db.get_storage_root(self._address)
+        if storage_root == BLANK_ROOT_HASH:
+            return HexaryTrie({})
+        elif self._write_trie is not None:
+            return self._write_trie
+        else:
+            if self._starting_root_hash is None:
+                self._starting_root_hash = storage_root
+            # Creating "HexaryTrie" is a pretty light operation, so not a huge cost
+            # to create a new one at every read, but we could
+            # cache the read trie, if this becomes a bottleneck.
+            return HexaryTrie(self._db, root_hash=storage_root)
+
+    def _decode_key(self, key: bytes) -> bytes:
+        padded_slot = pad32(key)
+        return keccak(padded_slot)
+
+    def __getitem__(self, key: bytes) -> bytes:
+        hashed_slot = self._decode_key(key)
+        read_trie = self._get_read_trie()
+        try:
+            return read_trie[hashed_slot]
+        except trie_exceptions.MissingTrieNode as exc:
+            self.logger.warning(exc)
+            self.logger.debug("expected to get slot, but:", exc_info=True)
+            raise exc
+
+    def __setitem__(self, key: bytes, value: bytes) -> None:
+        hashed_slot = self._decode_key(key)
+        write_trie = self._get_write_trie()
+        write_trie[hashed_slot] = value
+
+    def _exists(self, key: bytes) -> bool:
+        # used by BaseDB for __contains__ checks
+        hashed_slot = self._decode_key(key)
+        read_trie = self._get_read_trie()
+        return hashed_slot in read_trie
+
+    def __delitem__(self, key: bytes) -> None:
+        hashed_slot = self._decode_key(key)
+        write_trie = self._get_write_trie()
+        try:
+            del write_trie[hashed_slot]
+        except trie_exceptions.MissingTrieNode as exc:
+            self.logger.warning(exc)
+            self.logger.debug("expected to be able to delete slot, but:", exc_info=True)
+
+    @property
+    def has_changed_root(self) -> bool:
+        return self._write_trie and self._write_trie.root_hash != self._starting_root_hash
+
+    def get_changed_root(self) -> Hash32:
+        if self._write_trie is not None:
+            return self._write_trie.root_hash
+        else:
+            raise ValidationError("Asked for changed root when no writes have been made")
+
+    def clear_changed_root(self) -> None:
+        self._write_trie = None
+        self._starting_root_hash = None
```

Comment:
This still leaves data behind if the caller had already made changes, here's a failing test:

```python
def test_storage_lookup_clear_clears_changes(account_db):
    base_db = AtomicDB()
    lookup = StorageLookup(base_db, account_db, ADDRESS)

    target = dict()

    lookup[b'key'] = b'val' 
    lookup.clear_changed_root()
    assert not lookup.has_changed_root

    lookup.commit_to(target)
    assert target == {}  # fails here, because _trie_nodes_batch wasn't cleared by clear_changed_root
```

It might be a simpler user-interface if this method was dropped and `commit_to()` set these fields to None.

Reply:
Hah, yeah, I same to the same conclusion while handling the other comment above.